### PR TITLE
[NeMo-UX] Don't create attention mask for GPTs

### PR DESCRIPTION
**Pull request by JimmyZhang12**
# What does this PR do ?
For flash and fused attention kernels, the attention mask is implicitly generated by the kernel, so no need to generate the mask from the dataloader. This saves memory, especially for causal masks of long sequence lengths.

Flash attention and fused attention are well supported and should cover all the models supported by NeMo, and are already turned on by default, so turning this on by default as well. 

This also removes a HtoD at the start of the first layer, since the attention mask is on CPU for GPTDataset. This gives a ~5% speedup on the first pipeline parallel stage:
image

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by ShriyaPalsamudram — DISMISSED** [review]

File: nemo/collections/llm/gpt/data/pre_training.py [hunk 1]
```diff
@@ -11,6 +11,12 @@
 from nemo.lightning.io.mixin import IOMixin
 from nemo.lightning.pytorch.plugins import MegatronDataSampler
 
+HAVE_TE = True
+try:
+    import transformer_engine
```
[the remark this answers is not in the document]
Comment (JimmyZhang12) [reply nemo/collections/llm/gpt/data/pre_training.py:16]: This is ok as we just need to test that TE can be imported

File: nemo/collections/llm/gpt/data/mock.py [hunk 2]
```diff
@@ -9,6 +9,12 @@
 
 from nemo.lightning.pytorch.plugins import MegatronDataSampler
 
+HAVE_TE = True
+try:
+    import transformer_engine
+except (ImportError, ModuleNotFoundError):
```
Comment (ShriyaPalsamudram) [inline nemo/collections/llm/gpt/data/mock.py:15]: If the TE import fails, then the import error gets hidden in this except block. 

@marcromeyn @ericharper could you please share the right way to "import guard" in NeMo 2.0

Comment (JimmyZhang12) [reply nemo/collections/llm/gpt/data/mock.py:15]: I agreed with @ashors1 that we can first merge this with the messy imports and later refactor them when we have better import guards.

**Review by ShriyaPalsamudram — APPROVED** [review]
